### PR TITLE
TFLu: Port squeeze op from TFLite

### PR DESCRIPTION
**Pull request by mansnils**
This is a fix for: https://github.com/tensorflow/tensorflow/issues/44789

### CODE REVIEW

File: tensorflow/lite/micro/kernels/squeeze.cc [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  int input_num_dims = NumDimensions(op_context.input);
+  int num_squeeze_dims = op_context.params->num_squeeze_dims;
+
+  // Determines number of dimensions of output tensor after squeeze.
+  const TfLiteIntArray* input_dims = op_context.input->dims;
+  const int* squeeze_dims = op_context.params->squeeze_dims;
+  TF_LITE_ENSURE(context, input_num_dims <= 8);
+  bool should_squeeze[8] = {false};
+  int num_squeezed_dims = 0;
+  if (num_squeeze_dims == 0) {
```
Comment (advaitjain) [inline tensorflow/lite/micro/kernels/squeeze.cc:53]: this too is existing code, but it is really unfortunate naming.

I don't have a good suggestion at this time but calling this out that we need to change the names here.

File: tensorflow/lite/micro/kernels/squeeze.cc [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    }
+  } else {
+    for (int idx = 0; idx < num_squeeze_dims; ++idx) {
+      int current = squeeze_dims[idx] < 0 ? squeeze_dims[idx] + input_num_dims
+                                          : squeeze_dims[idx];
+      TF_LITE_ENSURE(context, current >= 0 && current < input_num_dims &&
+                                  input_dims->data[current] == 1);
+      if (!should_squeeze[current]) ++num_squeezed_dims;
+      should_squeeze[current] = true;
+    }
```
Comment (advaitjain) [inline tensorflow/lite/micro/kernels/squeeze.cc:68]: its not clear to me that this logic is actually doing something useful. In the TfLite implementation the should_squeeze array is being used, but that is not the case here:

https://github.com/tensorflow/tensorflow/blob/264780c8cf5fd06d11ad673bf351aa8b6850b974/tensorflow/lite/kernels/squeeze.cc#L51-L77

**Review by advaitjain — COMMENTED** [review]
I think we might be better off waiting until @petewarden can put together the Op porting guide.

One of the items there is breaking down the port into smaller PRs so that we can see the differences between the Lite and Micro implementation more easily.

**Comment by advaitjain** [comment]
Again, no changes needed until we hear from @petewarden but in the case there might be value in refactor to add a reference implementation of squeeze in [lite/kernels/internal/reference](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/kernels/internal/reference) and then have the Lite and Micro kernels make use of that shared code.

I haven't looked at this very carefully so very real chance of this comment being inaccurate, but it looks like the logic in the Prepare function might be shareable.

File: tensorflow/lite/micro/kernels/squeeze.cc [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+  SqueezeContext op_context(context, node);
+  int input_num_dims = NumDimensions(op_context.input);
+  int num_squeeze_dims = op_context.params->num_squeeze_dims;
+
+  // Determines number of dimensions of output tensor after squeeze.
+  const TfLiteIntArray* input_dims = op_context.input->dims;
+  const int* squeeze_dims = op_context.params->squeeze_dims;
+  TF_LITE_ENSURE(context, input_num_dims <= 8);
+  bool should_squeeze[8] = {false};
```
[the remark this answers is not in the document]
Comment (mansnils) [reply tensorflow/lite/micro/kernels/squeeze.cc:51]: Ok

File: tensorflow/lite/micro/kernels/squeeze.cc (hunk 1 above)
Comment (mansnils) [reply tensorflow/lite/micro/kernels/squeeze.cc:53]: At least num_squeezed_dims can be removed.

File: tensorflow/lite/micro/kernels/squeeze.cc (hunk 2 above)
Comment (mansnils) [reply tensorflow/lite/micro/kernels/squeeze.cc:68]: Good point! Let's just use it to check for valid dimensions instead.

**Comment by mansnils** [comment]
As discussed, let's continue with this PR.

**Comment by gbaned** [comment]
@mansnils  Can you please resolve conflicts? Thanks!

**Comment by mansnils** [comment]
@petewarden @advaitjain ping for review

**Review by petewarden — APPROVED** [review]

**Comment by mansnils** [comment]
@petewarden ping for re-review after resolving conflicts

**Comment by gbaned** [comment]
@mansnils  Can you please resolve conflicts? Thanks!

**Comment by mansnils** [comment]
@petewarden @advaitjain ping for review

**Comment by mansnils** [comment]
@petewarden @advaitjain ping for review

**Review by petewarden — APPROVED** [review]

**Comment by mansnils** [comment]
@gbaned ping for merge

**Comment by mansnils** [comment]
@gbaned ready for merge

**Comment by mansnils** [comment]
@advaitjain is there problem merging?

**Review by petewarden — APPROVED** [review]

**Comment by rthadur** [comment]
@mansnils can you please sync your branch to master ? thank you

**Comment by mansnils** [comment]
@rthadur branch is synced to master
@petewarden can you please re-approve? thanks

**Review by rthadur — APPROVED** [review]

**Pull request merged**